# Fundamentals: Python P2

## Warmup: adder function

Create a function that takes two numbers as arguments and return their sum.

### Examples

- `addition(3, 2) ➞ 5`
- `addition(-3, -6) ➞ -9`
- `addition(7, 3) ➞ 10`

In [3]:
# write your function here
def print_numbers(x,y):
    New_xy = x + y
    return New_xy
print_numbers(2,3)

5

In [8]:
# Warmup 2: even numbers
# Generate a Python list of all the even numbers between 4 to 3048
def evenNum():
    L = list()
    for num in range(4,3049):
        if num % 2==0:
            L.append(num)
    return L
evenNum()

[4,
 6,
 8,
 10,
 12,
 14,
 16,
 18,
 20,
 22,
 24,
 26,
 28,
 30,
 32,
 34,
 36,
 38,
 40,
 42,
 44,
 46,
 48,
 50,
 52,
 54,
 56,
 58,
 60,
 62,
 64,
 66,
 68,
 70,
 72,
 74,
 76,
 78,
 80,
 82,
 84,
 86,
 88,
 90,
 92,
 94,
 96,
 98,
 100,
 102,
 104,
 106,
 108,
 110,
 112,
 114,
 116,
 118,
 120,
 122,
 124,
 126,
 128,
 130,
 132,
 134,
 136,
 138,
 140,
 142,
 144,
 146,
 148,
 150,
 152,
 154,
 156,
 158,
 160,
 162,
 164,
 166,
 168,
 170,
 172,
 174,
 176,
 178,
 180,
 182,
 184,
 186,
 188,
 190,
 192,
 194,
 196,
 198,
 200,
 202,
 204,
 206,
 208,
 210,
 212,
 214,
 216,
 218,
 220,
 222,
 224,
 226,
 228,
 230,
 232,
 234,
 236,
 238,
 240,
 242,
 244,
 246,
 248,
 250,
 252,
 254,
 256,
 258,
 260,
 262,
 264,
 266,
 268,
 270,
 272,
 274,
 276,
 278,
 280,
 282,
 284,
 286,
 288,
 290,
 292,
 294,
 296,
 298,
 300,
 302,
 304,
 306,
 308,
 310,
 312,
 314,
 316,
 318,
 320,
 322,
 324,
 326,
 328,
 330,
 332,
 334,
 336,
 338,
 340,
 342,
 344,
 346,
 348,
 350,
 352,
 

## Exercise 1: First and Last 

Creates a function that takes a string and returns the first and last character.
### Examples

```
first_last("ganesh") ➞ "gh"

first_last("kali") ➞ "ki"

first_last("shiva") ➞ "sa"

first_last("vishnu") ➞ "vu"

first_last("durga") ➞ "da"
```

In [37]:
# exercise 1 --- i used a function that only allows one entry, it was my understanding that the question is 
# asking for a single string.
def first_last(x):
    return x[0]+x[-1]
print(first_last("ganesh"))
print(first_last("kali"))
print(first_last("shiva"))
print(first_last("vishnu"))
print(first_last("durga"))

gh
ki
sa
vu
da


## Exercise 2: Keys in Dict

Write a function that returns `True` if a dictionary contains the specified key, and `False` otherwise.

### Examples

```
has_key({ "a": 44, "b": 45, "c": 46 }, "d") ➞ False

has_key({ "craves": True, "midnight": True, "snack": True }, "morning") ➞ False

has_key({ "pot": 1, "tot": 2, "not": 3 }, "not") ➞ True
```

In [41]:
# exercise 2
def has_key (dict,key):
    if key in dict.keys():
        return("True")
    else:
        return("False")
    
print(has_key({ "a": 44, "b": 45, "c": 46},"c"))

True


## Exercise 3: Element Count

Write a function that count how many times each element in a list occurs. It should return a dict in the format `element -> count`

ex:
```
key_count([1, 3, 2, 1, 5, 3, 5, 1, 4]) ⇒ {
    1: 3,
    2: 1,
    3: 2,
    4: 1,
    5: 2,
}
```

In [21]:
# exercise 3 

def key_count(List):
    freq = {}
    for item in List:
        if item in freq:
            freq[item] += 1
        else: 
            freq[item] = 1 
    for key, value in freq.items():
        return freq
key_count([1, 3, 2, 1, 5, 3, 5, 1, 4])

{1: 3, 3: 2, 2: 1, 5: 2, 4: 1}

## Exercise 4: Test Distinct

Write a Python function that takes a sequence of numbers and determines whether all the numbers are different from each other.

```
test_distinct([2,3,4,5,7,8]) -> True
test_distinct([2,3,4,5,8,8]) -> False
```

In [58]:
# exercise 4
def test_distinct(x):
    if len(x) == len(set(x)) :
        print("True")
    if len(x) != len(set(x)):
        print("False")
test_distinct([2,3,4,5,7,8])
test_distinct([2,3,4,5,8,8])

True
False


## Exercise 5: Debt Calculator

Write a Python program to compute the amount of the debt in n months. The borrowing amount is $100,000 and the loan adds 5% interest of the debt and rounds it to the nearest 1,000 above month by month.

The function takes in an integer (the number of months) and returns the total debt including the starting \$100,000 rounded to \$1000 .

```
debt_total(7) -> 144000
debt_total(8) -> 152000
debt_total(9) -> 160000
debt_total(1) -> 105000
debt_total(57) -> 1736000
```


In [181]:
import math
def debt_total(n):
    PV = 100000
    x = 1
    while x <= n:
        PV = PV*1.05
        PV = int(math.ceil(PV/1000.0))*1000
        x += 1
    return PV
print(debt_total(7))
print(debt_total(8))
print(debt_total(9))
print(debt_total(1))
print(debt_total(57))

144000
152000
160000
105000
1736000


## Exercise 6: Spell Checker

Create a function that takes in two lists: 

1. A sentence as a list of user-typed words

2. The list of correctly-typed words

and outputs a list containing 1s (correctly-typed words) and 0s (incorrectly-typed words).

### Inputs:
> User-typed: \["cat", "blue", "skt", "umbrells", "paddy"\]

> Correct: \["cat", "blue", "sky", "umbrella", "paddy"\]

### Output: \[ 1, 1, 0, 0, 1 \]

### Examples

```
correct_stream(
  ["it", "is", "find"],
  ["it", "is", "fine"]
) ➞ [1, 1, 0]

correct_stream(
  ["april", "showrs", "bring", "may", "flowers"],
  ["april", "showers", "bring", "may", "flowers"]
) ➞ [1, 0, 1, 1, 1]

```

In [10]:
# exercise 6
def spell_check(a,b):
    spellCheck=[]
    for item in a:
        if item in b:
            spellCheck.append(1)
    else:
            spellCheck.append(0)
    return spellCheck
spell_check(["it", "is", "find"],["it", "is", "fine"])

[1, 1, 0]

## Exercise 7: FizzBuzz

Fizzbuzz is a game for children and a drinking game for adults. It's also a classic programming interview question.

Suppose we have a number n. We have to display a string representation of all numbers from 1 to n, but there are some constraints.

    If the number is divisible by 3, write Fizz instead of the number
    If the number is divisible by 5, write Buzz instead of the number
    If the number is divisible by 3 and 5 both, write FizzBuzz instead of the number

So the results are `1, 2, Fizz, 4, Buzz, ..., 14, Fizzbuzz, 16, 17, Fizz, 19, Buzz, ...`

Write a python program that prints fizzbuzz for numbers 1 to 150

In [145]:
# exercise 7
def fizzbuzz (x):
    x = range(1,151)
    FB = []
    for item in x:
        if item %3 == 0 and item %5 == 0:
            FB.append("FizzBuzz")
        elif item %5 == 0:
            FB.append("Buzz")
        elif item %3 == 0:
            FB.append("Fizz")
        else:
            FB.append(item)
    return(FB)
fizzbuzz(x)

[1,
 2,
 'Fizz',
 4,
 'Buzz',
 'Fizz',
 7,
 8,
 'Fizz',
 'Buzz',
 11,
 'Fizz',
 13,
 14,
 'FizzBuzz',
 16,
 17,
 'Fizz',
 19,
 'Buzz',
 'Fizz',
 22,
 23,
 'Fizz',
 'Buzz',
 26,
 'Fizz',
 28,
 29,
 'FizzBuzz',
 31,
 32,
 'Fizz',
 34,
 'Buzz',
 'Fizz',
 37,
 38,
 'Fizz',
 'Buzz',
 41,
 'Fizz',
 43,
 44,
 'FizzBuzz',
 46,
 47,
 'Fizz',
 49,
 'Buzz',
 'Fizz',
 52,
 53,
 'Fizz',
 'Buzz',
 56,
 'Fizz',
 58,
 59,
 'FizzBuzz',
 61,
 62,
 'Fizz',
 64,
 'Buzz',
 'Fizz',
 67,
 68,
 'Fizz',
 'Buzz',
 71,
 'Fizz',
 73,
 74,
 'FizzBuzz',
 76,
 77,
 'Fizz',
 79,
 'Buzz',
 'Fizz',
 82,
 83,
 'Fizz',
 'Buzz',
 86,
 'Fizz',
 88,
 89,
 'FizzBuzz',
 91,
 92,
 'Fizz',
 94,
 'Buzz',
 'Fizz',
 97,
 98,
 'Fizz',
 'Buzz',
 101,
 'Fizz',
 103,
 104,
 'FizzBuzz',
 106,
 107,
 'Fizz',
 109,
 'Buzz',
 'Fizz',
 112,
 113,
 'Fizz',
 'Buzz',
 116,
 'Fizz',
 118,
 119,
 'FizzBuzz',
 121,
 122,
 'Fizz',
 124,
 'Buzz',
 'Fizz',
 127,
 128,
 'Fizz',
 'Buzz',
 131,
 'Fizz',
 133,
 134,
 'FizzBuzz',
 136,
 137,
 'Fizz',
 1

## Exercise 8: Palindrome?

A palindrome is a word which reads the same backward as forwards. Write a function that returns True if a string is a palindrome:

    “laval” ⇒ Yes
    “montreal” ⇒ No

In [19]:
# exercise 8
def palindrome(x):
        if (x==x[::-1]):
            print(x,"yes")
        else:
            print(x,"no")
palindrome('laval')
palindrome('montreal')

laval yes
montreal no


## Exercise 9: Counting Country IDs

We have a list of user IDS with identifiers of form "userid-COUNTRYID". Calculate how many countries we have.

Ex:

`[gyhwefy7832ryyg-US, iu7ytf87gyuwef-CA, 98yfyg82t3r-US]` -> `{'US': 2, 'CA': 1}`

```
[gyhwefy7832ryyg-US, iu7ytf87gyuwef-CA, 98yfyg82t3r-US,
    iuhsdf897y3righsfe-US, iushdf98h23ryg-IT, iuhdf9h32ryu9gf-IT,
    IUufgh98y34-FR, 98yfdiuhdfbviuh8-UK
] -> {'US': 3, 'CA': 1, "IT": 2, 'UK':1, 'FR':1}
```

In [382]:
import collections
def countries(x):
    for country in (x):
        country =[country[-2:] for country in x]
        count = collections.Counter(country)
        return (count)
countries(["gyhwefy7832ryyg-US", "iu7ytf87gyuwef-CA", "98yfyg82t3r-US",
    "iuhsdf897y3righsfe-US", "iushdf98h23ryg-IT", "iuhdf9h32ryu9gf-IT",
    "IUufgh98y34-FR", "98yfdiuhdfbviuh8-UK"])

Counter({'US': 3, 'CA': 1, 'IT': 2, 'FR': 1, 'UK': 1})

## Exercise 10: Docstrings

Write the completed docstrings for each function covered in the "Big O Notation Examples" Section from lecture.

Docstring format is as follows:

```
"""[Summary]

:param [ParamName]: [ParamDescription], defaults to [DefaultParamVal]
:type [ParamName]: [ParamType](, optional)
...
:raises [ErrorType]: [ErrorDescription]
...
:return: [ReturnDescription]
:rtype: [ReturnType]
"""
```

In [106]:
# O(1)    check polynomial example in doc.string lecture. it's not a function in the lecture.
def O(x):
    """
    this function checks if argument is an even number/
    
    :param[items]: argument to be verified
    :type[items]: int
    ...
    :raises [ErrorType]: error arises if object is NOT float, int or boolean. EX: string won't work.
    ...
    :return: divides the argument by 2 and checks that remainder is == 0.
    :rtype: boolean
    """
    return x%2==0
print(O(9))
print(O(8))

False
True


In [96]:
# O(n) ------------------------ # i changed the function a little bit. i added 
def linear(items):
    """
    this function prints items from the argument
    
    :param[items]: elements to be printed
    :type[items]: object(can be string, list, int..ect)
    
    ...
    :return: displays the elements in argument
    :rtype: NoneType 
    """
## concerning 'rtype': since there is no return statment, the function doesn't return anything. 
#If I would write 'return' instead of 'print' then the 'rtype' would be the same as the argument type
    for item in items:
        print(item)
linear([4,5,6,8])

4
5
6
8


NoneType

In [57]:
# O(n^2)
def quadratic(items):
    """
    this function loops through all possible couple combinations from argument 
    
    :param[item]: elements picked to be coupled off 
    :type[item]:  iterable object(s)(ex: list,strings...ect)
    
    ...
    :return: the couples formed
    :rtype: NoneType - #default because no return statement
    """
    for item in items:
        for item2 in items:
            print(item,' ',item2)
    
quadratic([4,5,6,8])

4   4
4   5
4   6
4   8
5   4
5   5
5   6
5   8
6   4
6   5
6   6
6   8
8   4
8   5
8   6
8   8


In [107]:
# O(n^c)  
result = []
nums = [-1,0,1,2,-1,-4]

for i in range(len(nums)):
    for j in range(len(nums)):
        for k in range(len(nums)):
            if nums[i]+nums[j]+nums[k] == 0 and i!=j and i !=k and j !=k:
                x = sorted([nums[i], nums[j], nums[k]])
                if x not in result:
                    result.append(x)
"""
this loop defines 3 elements from the list, makes sure their total sum is "0" and that they are unique.
it then appends possible combinations into empty list: 'results' 

:param[nums]: these loop defines 3 elements from the list, makes sure their total sum is "0" and that they are unique.
it then appends possible combinations into empty list: 'results'
:type[nums]: List of int

...
:return: results of possible combinations
:rtype: list of int
"""
print(result)

[[-1, 0, 1], [-1, -1, 2]]


In [151]:
# O(log(n))

def binary_search(array,element):
    """
    Return the index of a given number within an array
    
    :param[array,element]: it checks if element = mid value(in length of array), if larger than cuts first half of list.
    repeats the process, keeps cutting array in half till left with element and its index.
    :type[array, element]:array, element
    
    ...
    :return: index of element
    :rtype: int
    """

    mid = 0 
    start = 0
    end = len(array)
    step = 0

    while (start <= end):
        step += 1
        mid = (start + end) // 2

        if element == array[mid]:
            return mid

        if element < array[mid]:
            end = mid - 1

        else:
            start = mid + 1

    return -1

element = 18

array = [1,2,5,7,13,15,16,18,20,28,29]

binary_search(array, element)

7

In [129]:
# O(nlog(n))

def mergeSort(X):
  '''
  sorting lists 
  
  :param[X]: the function goes through the list, sorts from the left and to the right,merges the results and moves 
  forward in the list repeating the process. after some accumulation, it goes back to the begining, sorts the build up
  (using the same process, going through all the objects). Does that till it gets to the end. 
  the final step is checking if any objects were missed in the merge, by verifying length of both lists are equivalent. 
  
  :type[X]:container of iterable objects. #the lecture example specifies a list of numbers, but the code would work 
                                          #with lists of strings or booleans. 
  ...
  :raises [ErrorType]: argument needs to be a container with >1 object. 
  ...
  :return: sorted argument
  :rtype: same container as input. 
  '''

  if len(X) > 1:
    mid = len(X)//2
 
    L = X[:mid]
    R = X[mid:]
    
    #Sort L
    mergeSort(L)
    #Sort R
    mergeSort(R)
    
    i = j = k = 0
    
    # add data to temp arrays L and R
    while i < len(L) and j < len(R):
        
        if L[i] < R[j]:
            X[k] = L[i]
            i += 1
        else:
            X[k] = R[j]
            j += 1
        k += 1
    
    #Last check to see if any items were left
    while i < len(L):
        X[k] = L[i]
        i += 1
        k += 1
    while j < len(R):
        X[k] = R[j]
        j += 1
        k += 1
    return(X)

num = [1,5,2,7,3,8]
mergeSort(num)

[1, 2, 3, 5, 7, 8]

In [144]:
# O(2^n)

def fibonacci(n):
    '''
    finds fibonacci equivalent to n. n considered as index within fibonacci sequence. 
  
    :param[n]: the code finds the two previous entries ((n-1)&(n-2)) and their fib equivalent, finally their sum.
    sum = fib equivalent of n
    :type[n]: int
    ...
    :raises [ErrorType]: argument has to be positive. 
    ...
    :return: fibonacci equivalent of n
    :rtype: int  
    '''
    if n <= 1:
        return n
    return fibonacci(n-1) + fibonacci(n-2)

fibonacci(10)

55

## Exercise 11: Big O Notation

For the following functions, answer in the text cell below, the time complexity of the function using Big O notation.

Ex.

```
def findmin(L):
 min = L[0]
 for number in L:
  if number<min:
    min = number 
```

The function has O(n) time complexity.

In [11]:
# 11.1
result = []
nums = [-1,0,1,2,-1,-4] 

for i in range(len(nums)):
    for j in range(len(nums)):
        for k in range(len(nums)):
            if nums[i]+nums[j]+nums[k] == 0 and i!=j and i !=k and j !=k:
                x = sorted([nums[i], nums[j], nums[k]])
                if x not in result:
                    result.append(x)

Answer: O(n^c)

In [58]:
# 11.2 
def add_hundred(n):
    return n + 100

# n can only be an integer or float.

Answer: O(C)